lib/srv/db/redis/client.go

```diff
@@ -0,0 +1,287 @@
+//   then it's handled by Teleport or blocked if reasonable processing is not possible.
+// * Otherwise a commands is sent to Redis without any modifications.
+func (c *clusterClient) Process(ctx context.Context, inCmd redis.Cmder) error {
+	cmd, ok := inCmd.(*redis.Cmd)
+	if !ok {
+		return trace.BadParameter("failed to cast Redis command type: %T", cmd)
+	}
+
+	switch cmdName := strings.ToLower(cmd.Name()); cmdName {
+	case multiCmd, execCmd, watchCmd, scanCmd, aclCmd, askingCmd, clientCmd, clusterCmd, configCmd, debugCmd,
+		infoCmd, latencyCmd, memoryCmd, migrateCmd, moduleCmd, monitorCmd, pfdebugCmd, pfselftestCmd,
+		psyncCmd, readonlyCmd, readwriteCmd, replconfCmd, replicaofCmd, roleCmd, shutdownCmd, slaveofCmd,
+		slowlogCmd, syncCmd, timeCmd, waitCmd:
+		// block commands that return incorrect results in Cluster mode
+		return trace.NotImplemented("%s is not supported in the cluster mode", cmdName)
+	case dbsizeCmd:
+		// use go-redis dbsize implementation. It returns size of all keys in the whole cluster instead of
+		// just currently connected instance.
+		res := c.DBSize(ctx)
+		cmd.SetVal(res.Val())
+		cmd.SetErr(res.Err())
+
+		return nil
+	case keysCmd:
+		var resultsKeys []string
+		var mtx sync.Mutex
+
+		if err := c.ForEachMaster(ctx, func(ctx context.Context, client *redis.Client) error {
+			scanCmd := redis.NewStringSliceCmd(ctx, cmd.Args()...)
+			err := client.Process(ctx, scanCmd)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			keys, err := scanCmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			mtx.Lock()
+			resultsKeys = append(resultsKeys, keys...)
+			mtx.Unlock()
+
+			return nil
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+
+		cmd.SetVal(resultsKeys)
+
+		return nil
+	case mgetCmd:
+		if len(cmd.Args()) == 1 {
+			return trace.BadParameter("wrong number of arguments for 'mget' command")
+		}
+
+		var resultsKeys []interface{}
+
+		keys := cmd.Args()[1:]
+		for _, k := range keys {
+			result := c.Get(ctx, k.(string))
+			if result.Err() == redis.Nil {
+				resultsKeys = append(resultsKeys, redis.Nil)
+				continue
+			}
+
+			if result.Err() != nil {
+				cmd.SetErr(result.Err())
+				return trace.Wrap(result.Err())
+			}
+
+			resultsKeys = append(resultsKeys, result.Val())
+		}
+
+		cmd.SetVal(resultsKeys)
+
+		return nil
+	case flushallCmd, flushdbCmd:
+		var mtx sync.Mutex
+
+		if err := c.ForEachMaster(ctx, func(ctx context.Context, client *redis.Client) error {
+			singleCmd := redis.NewCmd(ctx, cmd.Args()...)
+			err := client.Process(ctx, singleCmd)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			mtx.Lock()
+			defer mtx.Unlock()
+
+			if cmd.Err() != nil {
+				// If other call have already set error do not erase it.
+				// It should be returned to the caller.
+				return nil
+			}
+
+			cmd.SetVal(singleCmd.Val())
```

Comment:
Wonder if in case where the flow is skipped by the `continue` call  when `cmd.Err() != nil` the cmd.Val and cmd.Err can be both `nil`. 
In that case `processSeverResponse` can return `nil, nil` so further  flow can panic.